### ISSUE TITLE
stackwalk.cpp assumes locals are 16-byte aligned

### ISSUE THREAD

**Issue by kangaroo** [dotnet/coreclr#141]
The implementation of moveOWord currently uses movdqa on all platforms, meaning that src and target both need to be 16 byte aligned.  Currently the only caller is atomicMoveCacheEntry, which is called from  StackwalkCache::Insert.

StackwalkCache::Insert is called like so:
```
                StackwalkCacheEntry cacheEntry = {0};                                                                                                                                                                                                 
                if (cacheEntry.Init(                                                  
                            curIP,                                                    
                            SPOffset,                                                 
                            &unwindInfo,                                              
                            argSize))                                                 
                {                                                                     
                    m_crawl.stackWalkCache.Insert(&cacheEntry);                       
                }     
```

As such a stack pointer is passed directly down to the helper.  On OS X at least, this pointer is only guaranteed 8 byte alignment.

We can fix this with  __attribute__((aligned(16))), but thats non-portable.  Should we start to introduce a platform agnostic #define to cover alignment needs?

**Comment by sergiy-k** [comment]
@kangaroo The declaration of the StackwalCacheEntry class already has DECLSPEC_ALIGN(STACKWALK_CACHE_ENTRY_ALIGN_BOUNDARY) which is defined in pal.h. Maybe it does not work with llvm on OS X if you are hitting this problem at runtime?

**Comment by kangaroo** [comment]
I was definitely hitting this at runtime, but I see the decl on the entry now, and it seems to be pulling in the right thing from pal.h

I'll do some additional digging into why this is happening.

**Comment by sergiy-k** [comment]
By the way, things like stack walking and exception handling have not been yet implemented for Unix platforms. We are currently working on them.

**Comment by jkotas** [comment]
The alignment is really only required on x86 because of the lock-free implementation. 

For x64, it should be fine to get rid of the alignment and moveOWord assembly implementation. It is not required because of atomicMoveCacheEntry takes a lock.

**Comment by kangaroo** [comment]
Here's the issue:

basalt:Work plasma$ clang++ -std=c++11 test.cpp -o t 
test.cpp:3:12: warning: attribute 'align' is ignored, place it after "struct" to apply attribute to type declaration [-Wignored-attributes]
__declspec(align(0x10))
           ^
1 warning generated.

It wasn't immediately obvious because we set -Wno-ignored-attributes by default, suppressing this message.

**Comment by janvorli** [comment]
I don't get this warning on Linux (even with the ignored-attributes warnings enabled). It looks like a bug in the Apple's clang, the __declspec is supposed to be placed in front of the struct declaration. It looks as if that version of clang just converts it to __attribute__ without considering that.

**Comment by janvorli** [comment]
I don't get this warning on Linux (even with the ignored-attributes warnings enabled). It looks like a bug in the Apple's clang, the \_\_declspec is supposed to be placed in front of the struct declaration. It looks as if that version of clang just converts it to \_\_attribute\_\_ without considering that.

**Comment by kangaroo** [comment]
@janvorli fixed this is #282.  Can someone close this out please?

**Comment by kangaroo** [comment]
This was fixed in #282.
